src/mbgl/gl/parameters.hpp

```diff
@@ -0,0 +1,234 @@
+    Range<float> range;
+
+    static Depth disabled() {
+       return Depth { Always, false, { 0.0, 1.0 } };
+    }
+};
+
+class Stencil {
+public:
+    template <GLenum>
+    struct SimpleTest {
+    };
+
+    template <GLenum>
+    struct MaskedTest {
+        GLuint mask;
+    };
+
+    using Always   = SimpleTest<GL_ALWAYS>;
+    using Never    = SimpleTest<GL_NEVER>;
+    using Less     = MaskedTest<GL_LESS>;
+    using LEqual   = MaskedTest<GL_LEQUAL>;
+    using Greater  = MaskedTest<GL_GREATER>;
+    using GEqual   = MaskedTest<GL_GEQUAL>;
+    using Equal    = MaskedTest<GL_EQUAL>;
+    using NotEqual = MaskedTest<GL_NOTEQUAL>;
+
+    using Test = variant<
+        Always,
+        Never,
+        Less,
+        LEqual,
+        Greater,
+        GEqual,
+        Equal,
+        NotEqual>;
+
+    enum Op {
+        Keep     = GL_KEEP,
+        Zero     = GL_ZERO,
+        Replace  = GL_REPLACE,
+        Incr     = GL_INCR,
+        IncrWrap = GL_INCR_WRAP,
+        Decr     = GL_DECR,
+        DecrWrap = GL_DECR_WRAP,
+        Inver    = GL_INVERT
+    };
+
+    Test test;
+    GLint ref;
+    GLuint mask;
+
+    Op fail;
+    Op depthFail;
+    Op pass;
+
+    static Stencil disabled() {
+       return Stencil { Always(), 0, 0, Keep, Keep, Keep };
+    }
+};
+
+class Blend {
+public:
+    enum Factor {
+        Zero                  = GL_ZERO,
+        One                   = GL_ONE,
+        SrcColor              = GL_SRC_COLOR,
+        OneMinusSrcColor      = GL_ONE_MINUS_SRC_COLOR,
+        DstColor              = GL_DST_COLOR,
+        OneMinusDstColor      = GL_ONE_MINUS_DST_COLOR,
+        SrcAlpha              = GL_SRC_ALPHA,
+        OneMinusSrcAlpha      = GL_ONE_MINUS_SRC_ALPHA,
+        DstAlpha              = GL_DST_ALPHA,
+        OneMinusDstAlpha      = GL_ONE_MINUS_DST_ALPHA,
+        ConstantColor         = GL_CONSTANT_COLOR,
+        OneMinusConstantColor = GL_ONE_MINUS_CONSTANT_COLOR,
+        ConstantAlpha         = GL_CONSTANT_ALPHA,
+        OneMinusConstantAlpha = GL_ONE_MINUS_CONSTANT_ALPHA,
+        SrcAlphaSaturate      = GL_SRC_ALPHA_SATURATE,
+    };
+
+    template <GLenum F>
+    struct ConstantFunction {
+    };
+
+    template <GLenum F>
+    struct LinearFunction {
+        Factor srcFactor;
+        Factor dstFactor;
+    };
+
+    struct Replace {};
+    using Min              = ConstantFunction<GL_MIN>;
+    using Max              = ConstantFunction<GL_MAX>;
+    using Add              = LinearFunction<GL_FUNC_ADD>;
+    using Subtract         = LinearFunction<GL_FUNC_SUBTRACT>;
+    using ReverseSubtract  = LinearFunction<GL_FUNC_REVERSE_SUBTRACT>;
+
+    using Function = variant<
+        Replace,
+        Min,
+        Max,
+        Add,
+        Subtract,
+        ReverseSubtract>;
+
+    Function function;
+    Color color;
+
+    static Blend disabled() {
+       return Blend { Replace(), {} };
+    }
+};
+
+class ColorMask {
+public:
+    bool r;
+    bool g;
+    bool b;
+    bool a;
```

Comment:
For types like this, we should either default them, or write explicit initializers that require all parameters to prevent accidental misuse, which results in uninitialized values. Valgrind may catch them, but it's always better to have a compiler error slapped in the face.